File: android/src/main/java/com/mobilecoin/mobilecoin_flutter/FfiAccountKey.java
// Copyright (c) 2021-2024 MobileCoin. All rights reserved.

package com.mobilecoin.mobilecoin_flutter;

import android.net.Uri;

import androidx.annotation.Keep;

import com.mobilecoin.lib.AccountKey;
import com.mobilecoin.lib.PublicAddress;
import com.mobilecoin.lib.exceptions.InvalidUriException;
import com.mobilecoin.lib.exceptions.BadEntropyException;

@Keep
public class FfiAccountKey {

    private FfiAccountKey() {
    }

    public static int fromBip39Entropy(byte[] entropy, String fogReportUri, String reportId, byte[] fogAuthoritySpki)
            throws InvalidUriException, BadEntropyException {
        AccountKey accountKey = AccountKey.fromBip39Entropy(entropy, 0, Uri.parse(fogReportUri), reportId,
                fogAuthoritySpki);
        final int hashCode = accountKey.hashCode();
        ObjectStorage.addObject(hashCode, accountKey);
        return hashCode;
    }

    public static int getPublicAddress(int addressId) {
        AccountKey accountKey = (AccountKey) ObjectStorage.objectForKey(addressId);
        PublicAddress publicAddress = accountKey.getPublicAddress();
        final int hashCode = publicAddress.hashCode();
        ObjectStorage.addObject(hashCode, publicAddress);
        return hashCode;
    }
}

File: ios/mobilecoin_flutter/Sources/mobilecoin_flutter/FfiAccountKey.swift
//
//  FfiAccountKey.swift
//  mobilecoin_flutter
//
//  Created by Alexander Voloshyn on 9/29/20.
//

import Flutter
import Foundation
import MobileCoin

struct FfiAccountKey {
    struct FromBip39Entropy: Command {
        func execute(args: [String: Any], result: @escaping FlutterResult) throws {
            guard let entropy = args["bip39Entropy"] as? FlutterStandardTypedData,
                  let fogReportUri: String = args["fogReportUri"] as? String,
                  let fogAuthoritySpki = args["fogAuthoritySpki"] as? FlutterStandardTypedData,
                  let reportId: String = args["reportId"] as? String else {
                      throw PluginError.invalidArguments
                  }
            let accountKey = AccountKey.make(entropy: entropy.data, fogReportUrl: fogReportUri, fogReportId: reportId, fogAuthoritySpki: fogAuthoritySpki.data,  accountIndex: 0)
            switch accountKey {
            case .success(let accountKey):
                let hash = accountKey.hashValue
                ObjectStorage.addObject(accountKey, forKey: hash)
                result(hash)
            case .failure(let error):
                throw error
            }
        }
    }
    struct GetPublicAddress: Command {
        func execute(args: [String : Any], result: @escaping FlutterResult) throws {
            guard let addressId: Int = args["id"] as? Int else {
                throw PluginError.invalidArguments
            }
            let accountKey: AccountKey = ObjectStorage.objectForKey(addressId) as! AccountKey
            let publicAddress: PublicAddress = accountKey.publicAddress
            let hashCode = publicAddress.hashValue
            ObjectStorage.addObject(publicAddress, forKey: hashCode)
            result(hashCode)
        }
    }
}

